### PR TITLE
DC/MLX5/OPTIMIZATION: use bitmap to release detached dci's

### PR DESCRIPTION
**Pull request by hoopoepg**
- use ucs_bitmap to store set of dci's to release instead of check every DCI

### CODE REVIEW

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 1]
```diff
@@ -1291,29 +1291,30 @@ uct_dc_mlx5_iface_dci_do_common_pending_tx(uct_dc_mlx5_ep_t *ep,
...
 {
     uct_dc_mlx5_iface_t *iface = arg;
-    size_t ndci                = iface->tx.ndci * iface->tx.num_dci_pools;
     uint64_t pool_map          = 0;
     uint8_t pool_index;
     uint8_t dci;
 
     ucs_assert(iface->tx.dci_release_prog_id != UCS_CALLBACKQ_ID_NULL);
     ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
+    UCS_STATIC_ASSERT(UCT_DC_MLX5_IFACE_MAX_DCIS <
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1300]: nice 👍

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 2]
```diff
@@ -483,7 +483,10 @@ uct_dc_mlx5_iface_dci_detach(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep,
 
     ep->dci                      = UCT_DC_MLX5_EP_NO_DCI;
     ep->flags                   &= ~UCT_DC_MLX5_EP_FLAG_TX_WAIT;
-    iface->tx.dcis[dci_index].ep = new_ep;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:486]: can't this create some inconsistency that ep ep no longer points to dci, but dci still points to ep?
or maybe add new function schedule_release which will just UCS_BITMAP_SET and add progress callback to release. and when we want to release the dci from purge context of pending context, just call schedule_release

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h:486]: ok, refactored

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h:486]: also added dci->ep=NULL

File: src/ucs/datastruct/bitmap.h [hunk 3]
```diff
@@ -45,6 +45,17 @@ typedef uint64_t ucs_bitmap_word_t;
 #define _UCS_BITMAP_NUM_WORDS(_bitmap) ucs_static_array_size((_bitmap).bits)
 
 
+/**
+ * Get the number of bits in a given bitmap
+ *
+ * @param _bitmap Bits number in this bitmap
```
Comment (brminich) [inline src/ucs/datastruct/bitmap.h:51]: it is just a bitmap itself, not a number of bits

**Review by brminich — APPROVED** [review]

File: src/ucs/datastruct/bitmap.h (hunk 3 above)
Comment (hoopoepg) [reply src/ucs/datastruct/bitmap.h]: reworded

**Review by brminich — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 4]
```diff
@@ -384,6 +385,7 @@ uct_dc_mlx5_iface_dci_release(uct_dc_mlx5_iface_t *iface, uint8_t dci_index)
 
     pool->stack_top--;
     pool->stack[pool->stack_top] = dci_index;
+    iface->tx.dcis[dci_index].ep = NULL;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:388]: is this still needed?

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h:388]: yes due to https://github.com/openucx/ucx/blob/ab202c97e5d61f166acb9538ab746a36b23bcbc5/src/uct/ib/dc/dc_mlx5_ep.h#L460

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.h:388]: but isn't setting it to NULL in uct_dc_mlx5_iface_dci_detach() enough?

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h:388]: no: dci_release is called from poll_tx->dci_put

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.h:388]: maybe restore 'iface->tx.dcis[dci_index].ep = NULL;' in dci_put?

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h]: done

**Review by yosefe — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Pull request merged**